mesonbuild/minstall.py

```diff
@@ -555,7 +555,8 @@ def do_install(self, datafilename: str) -> None:
                 if not self.options.quiet and self.preserved_file_count > 0:
                     self.log('Preserved {} unchanged files, see {} for the full list'
                              .format(self.preserved_file_count, os.path.normpath(self.lf.name)))
-        except PermissionError:
+        except PermissionError as e:
+            print(f'Failed to install {e.filename}')
```

Comment:
This will print twice if the handling code here says "ok don't both running a rootcmd and just reraise the exception".

Reply:
yes, but that error appears to be unhandled right now anyway? So I'd say that can go into a separate PR.

In the normal-ish case where one would cancel sudo/pkexec the error is handled correctly:

```
$ MESON_ROOT_CMD=pkexec ninja install
[1/2] Installing files.
Installing subdir /home/whot/code/libinput/quirks to /usr/share/libinput
Failed to install /usr/share/libinput/10-generic-keyboard.quirks     # the newly added line
Installation failed due to insufficient permissions.
Attempt to use pkexec to gain elevated privileges? [y/n] y
<hit escape/cancel>
Error executing command as another user: Request dismissed
FAILED: meson-internal__install 
/usr/bin/meson install --no-rebuild
ninja: build stopped: subcommand failed.
```

````
MESON_ROOT_CMD=sudo ninja install  
[1/2] Installing files.
Installing subdir /home/whot/code/libinput/quirks to /usr/share/libinput
Failed to install /usr/share/libinput/10-generic-keyboard.quirks      # the newly added line
Installation failed due to insufficient permissions.
Attempt to use sudo to gain elevated privileges? [y/n] y
[sudo] password for whot:
<Ctrl+C>
sudo: a password is required
ninja: build stopped: interrupted by user.
```

Reply:
> yes, but that error appears to be unhandled right now anyway? So I'd say that can go into a separate PR.

Dunno what you mean by unhandled. We literally handle it by reraising a well-formatted PermissionError that includes lots of context on what went wrong:
- the name of the file
- "permission denied"
- "probably an issue with your build environment" -- ok that one is a bit awkwardly wordy, but it's the fault of commit fa044e011d5d7a3525c9b3f2acfe1cbf4a0a96a1 and finding the right balance may be hard.

Reply:
> In the normal-ish case where one would cancel sudo/pkexec the error is handled correctly:

It actually isn't "handled correctly" but if we get as far as os.execlp then the entire context is destroyed and we literally cannot report anything else at that point. If a rootcmd does exist but we end up not using it e.g. because the user said "no" -- canceling sudo -- then we once again reraise.

Reply:
hehe, by "unhandled" I'm basically saying the same as meson tells me: `Unhandled python OSError.  This is probably not a Meson bug, but an issue with your build environment.`

And with the traceback, etc. the output looks *very* identical to an unhandled exception in python code otherwise (even though here it's re-raised and wrapped with the message but tbh for a while I had though it just crashed).
For something like "couldn't install the file because of permission issues" that seems like it could be more polished.

I think we're looking at this from different perspectives: from the code perspective, the error is handled when the rootcmd is declined and it's not handled in the `exceclp` case. But from a user interface perspective it looks like the error is unhandled in the rootcmd "no" case and handled in the `execlp` case.